addons/mail/wizard/mail_compose_message.py

```diff
@@ -512,17 +512,26 @@ def _process_state(self, mail_values_dict):
 
         return mail_values_dict
 
-    def _get_blacklist_record_ids(self, mail_values_dict):
+    def _get_blacklist_record_ids(self, mail_values_dict, recipients_info):
         blacklisted_rec_ids = set()
-        if self.composition_mode == 'mass_mail' and issubclass(type(self.env[self.model]), self.pool['mail.thread.blacklist']):
+        if self.composition_mode == 'mass_mail':
             self.env['mail.blacklist'].flush(['email'])
             self._cr.execute("SELECT email FROM mail_blacklist WHERE active=true")
             blacklist = {x[0] for x in self._cr.fetchall()}
             if blacklist:
-                targets = self.env[self.model].browse(mail_values_dict.keys()).read(['email_normalized'])
-                # First extract email from recipient before comparing with blacklist
-                blacklisted_rec_ids.update(target['id'] for target in targets
-                                           if target['email_normalized'] in blacklist)
+                if issubclass(type(self.env[self.model]), self.pool['mail.thread.blacklist']):
+                    targets = self.env[self.model].browse(mail_values_dict.keys()).read(['email_normalized'])
+                    # First extract email from recipient before comparing with blacklist
+                    blacklisted_rec_ids.update(target['id'] for target in targets
+                                               if target['email_normalized'] in blacklist)
+                else:
+                    # Note that we exclude the record if at least one recipient is blacklisted (-> even if not all)
+                    blacklisted_rec_ids.update(res_id for res_id, recipient_info in recipients_info.items()
+                                               if blacklist & set(recipient_info['mail_to_normalized']))
+                    for res_id, mail_value_dict in mail_values_dict.items():
+                        email_to = mail_value_dict.get('email_to')
+                        if email_to and tools.email_normalize(email_to) in blacklist:
+                            blacklisted_rec_ids.add(res_id)
```

Comment:
Can be onelined

```python
blacklisted_rec_ids |= {
    res_id for res_id, values in mail_values_dict.items()
    if values.get('email_to') and tools.email_normalize(values['email_to']) in blacklist
}
```

Reply:
Actually, I have realized that email_to is already included in "mail_to_normalized" by the method _process_recipient_values. So I have removed completely those statements.